File: cs/competitive_programming/leetcode_fulltimes/2023/januarary/cpp/49_group_anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {

        map<vector<int>, vector<string>> store;

        for(auto x: strs){
            vector<int> v(26,0);
            for(auto ch: x){
                v[ch-'a']++;
            }
            store[v].push_back(x);
        }

        vector<vector<string>> res;

        for(auto x: store){
            res.push_back(x.second);
        }

        return res;
    }
};


class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        std::unordered_map<std::string, int> map;
        std::vector<vector<string>> res;

        for (const auto &s: strs) {
            std::string sorted = s;
            std::sort(sorted.begin(), sorted.end());

            if (map.count(sorted)) {
                res[map[sorted]].push_back(s);
            } else {
                res.push_back({s});
                map[sorted] = res.size() - 1;
            }
        }

        return res;
    }
};


class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {

        vector<vector<string>> res;
        map<string, int> idx_finder;

        for(auto &x: strs){
            string tmp=x;
            sort(tmp.begin(), tmp.end());

            if(idx_finder.count(tmp)){
                res[idx_finder[tmp]].push_back(x);
            } else {
                res.push_back({x});
                idx_finder[tmp]=res.size()-1;
            }

        }

        return res;
        
    }
};

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        
        map<string, vector<string>> store;
        vector<vector<string>> res;


        for(auto &x: strs){
            string tmp = x;
            sort(x.begin(), x.end());
            store[x].push_back(tmp);
        }

        for(auto it=store.begin(); it!=store.end(); it++){
            res.push_back(it->second);
        }

        return res;


    }
};

File: cs/competitive_programming/leetcode_interns/OtherLeetCodeProblems/Facebook/Mock/49_group_anagrams.py
#!/usr/bin/python3


# https://leetcode.com/problems/group-anagrams/





from typing import DefaultDict


class Solution:
    def groupAnagrams(self, strs):

        if len(strs) == 1:
            return [strs]

        res = {}
        

        for ele in strs:

            lst = [0]*26
            
            for c in ele:

                val = ord(c) - ord('a')
                lst[val] += 1
            print(lst)
            lst = tuple(lst)
            print(lst)

            if lst in res.keys():
                res[lst].append(ele)
            else:
                res[lst] = [ele]
        
        return res.values()

class Solution1:
    def groupAnagrams(self, strs):

        if len(strs) == 1:
            return [strs]
            
        res = {}
        
        for ele in strs:
            tmp = list(ele)
            tmp.sort()
            tmp = ''.join(tmp)

            if tmp in res.keys():
                res[tmp].append(ele)
            else:
                res[tmp] = [ele]
        
        return res.values()


# obj = Solution()
# strs = ["eat","tea","tan","ate","nat","bat"]

# out = obj.groupAnagrams(strs)

# print(out)

obj = Solution1()
strs = ["eat","tea","tan","ate","nat","bat"]

out = obj.groupAnagrams(strs)

print(out)
